src/main/java/org/opensearch/neuralsearch/mappingtransformer/SemanticMappingTransformer.java

```
@@ -0,0 +1,425 @@
+                            "%s should not be longer than %d characters for " + "the semantic field at %s",
+                            SEMANTIC_INFO_FIELD_NAME,
+                            FIELD_NAME_MAX_LENGTH,
+                            semanticFieldPath
+                        )
+                    );
+                }
+            } else {
+                errors.add(
+                    String.format(
+                        Locale.ROOT,
+                        "%s should be a non-empty string for the semantic field" + " at %s",
+                        SEMANTIC_INFO_FIELD_NAME,
+                        semanticFieldPath
+                    )
+                );
+            }
+        }
+        return errors;
+    }
+
+    private void fetchModelAndModifyMapping(
+        @NonNull final Map<String, Map<String, Object>> semanticFieldPathToConfigMap,
+        @NonNull final Map<String, Object> mappings,
+        @NonNull final ActionListener<Void> listener
+    ) {
+        final Map<String, List<String>> modelIdToFieldPathMap = SemanticMappingUtils.extractModelIdToFieldPathMap(
+            semanticFieldPathToConfigMap
+        );
+        if (modelIdToFieldPathMap.isEmpty()) {
+            listener.onResponse(null);
+        }
+        final AtomicInteger counter = new AtomicInteger(modelIdToFieldPathMap.size());
+        final AtomicBoolean hasError = new AtomicBoolean(false);
+        final List<String> errors = new ArrayList<>();
+        final Map<String, MLModel> modelIdToConfigMap = new HashMap<>(modelIdToFieldPathMap.size());
+
+        // We can have multiple semantic fields with different model ids, and we should get model config for each model
+        for (String modelId : modelIdToFieldPathMap.keySet()) {
+            mlClientAccessor.getModel(modelId, ActionListener.wrap(mlModel -> {
+                modelIdToConfigMap.put(modelId, mlModel);
+                if (counter.decrementAndGet() == 0) {
```

why wait til last iteration for this check? can we fail fast as soon as error is encountered?

---

The API calls are executed in parallel so we have to wait for all done and then decide how to call back the listener. If we simply call back the listener then the listener can be invoked multiple times.